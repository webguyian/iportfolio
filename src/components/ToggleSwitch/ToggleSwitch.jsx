import React, { useState } from 'react';
import { func, string } from 'prop-types';
import classNames from 'classnames';

import Button from 'components/Button/Button';
import Text from 'components/Text/Text';

const ToggleSwitch = props => {
  const { label, onUpdate } = props;
  const [checked, setChecked] = useState(false);
  const baseClass = 'ui-toggle-switch';
  const activeClass = checked ? `${baseClass}--active` : false;

  const handleChange = ({ target }) => {
    setChecked(target.checked);
  };

  return (
    <label className={classNames(baseClass, activeClass)}>
      <input
        className={`${baseClass}-checkbox`}
        type="checkbox"
        checked={checked}
        onChange={handleChange}
      />
      <Button className={`${baseClass}-slider`} onTransitionEnd={onUpdate} />
      <Text className={`${baseClass}-label`}>{label}</Text>
    </label>
  );
};

ToggleSwitch.propTypes = {
  label: string,
  onUpdate: func.isRequired
};

ToggleSwitch.defaultProps = {
  label: 'slide to unlock'
};

export default ToggleSwitch;
