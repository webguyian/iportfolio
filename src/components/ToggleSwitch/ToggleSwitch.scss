.ui-toggle-switch {
  display: block;
  position: relative;
  margin: 0 auto;
  border-radius: 25px;
  background-color: rgba($black, 0.5);
  text-align: center;
  cursor: grab;

  &-checkbox {
    @include accessibly-hidden;
  }

  &-slider {
    display: block;
    position: absolute;
    top: 8px;
    left: 8px;
    width: 30px;
    height: 30px;
    transition: 0.3s transform;
    border-radius: 50%;
    background-color: $white;
    content: "";

    .ui-toggle-switch--active & {
      transform: translateX(170px);
    }
  }

  &-label {
    display: inline-block;
    padding: 10px 40px;
    transform: translateX(10px);
    transition: color 0.6s, transform 0.3s;
    color: $gray;
    font-size: $font-size-l;
    font-weight: $font-weight-light;

    .ui-toggle-switch--active & {
      transform: translateX(-10px);
      color: $white;
    }
  }
}
