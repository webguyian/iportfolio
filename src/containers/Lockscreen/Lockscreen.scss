/* stylelint-disable color-no-hex */
.iportfolio-lockscreen {
  @include flex-column;
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
  padding-bottom: 50px;
  transition: 0.4s transform cubic-bezier(0.25, 0.46, 0.45, 0.94);
  background: linear-gradient(320deg, #cdee23 0%, #74bf82 40%, $blue-light 60%, #0056ff 100%);

  &--unlocked {
    transform: translateX(110%);
  }

  &-header {
    display: flex;
    flex-flow: column wrap;
    margin-top: $gutter-l;

    .ui-clock {
      @include type-number;
      margin: 0;
    }

    .ui-date {
      margin-top: 0;
      font-size: $font-size-l;
      font-weight: $font-weight-light;
    }
  }
}
