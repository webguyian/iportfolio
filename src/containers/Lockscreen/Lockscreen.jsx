import React, { Component } from 'react';
import classNames from 'classnames';
import { Redirect } from 'react-router-dom';

import Icon from 'components/Icon/Icon';
import Text from 'components/Text/Text';
import ToggleSwitch from 'components/ToggleSwitch/ToggleSwitch';

class Lockscreen extends Component {
  constructor(props) {
    super(props);

    this.handleRedirect = this.handleRedirect.bind(this);
    this.handleUnlock = this.handleUnlock.bind(this);

    this.state = {
      redirect: false,
      unlocked: false
    };
  }

  handleRedirect() {
    this.setState(() => ({
      redirect: true
    }));
  }

  handleUnlock() {
    this.setState(() => ({
      unlocked: true
    }));
  }

  render() {
    const { redirect, unlocked } = this.state;
    const baseClass = 'iportfolio-lockscreen';
    const unlockedClass = unlocked && `${baseClass}--unlocked`;
    const dateObj = new Date();
    const [time] = dateObj
      .toLocaleString([], {
        hour: 'numeric',
        minute: '2-digit',
        hour12: true
      })
      .split(' ');

    const date = dateObj.toLocaleString([], {
      weekday: 'long',
      month: 'long',
      day: 'numeric'
    });

    return (
      <div
        className={classNames(baseClass, unlockedClass)}
        onTransitionEnd={this.handleRedirect}
      >
        <header className={`${baseClass}-header`}>
          <Icon name="lock" size="2x" />
          <Text className="ui-clock" element="h1" type="display">
            {time}
          </Text>
          <Text className="ui-date" element="h1" type="display">
            {date}
          </Text>
        </header>
        <ToggleSwitch onUpdate={this.handleUnlock} />
        {redirect && <Redirect to="/home" />}
      </div>
    );
  }
}

export default Lockscreen;
