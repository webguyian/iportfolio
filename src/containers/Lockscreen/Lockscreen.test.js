import React from 'react';
import { act, create } from 'react-test-renderer';

import Lockscreen from './Lockscreen';
import ToggleSwitch from 'components/ToggleSwitch/ToggleSwitch';

describe('<Lockscreen />', () => {
  const props = {
    history: { push: jest.fn() }
  };

  it('renders correctly', () => {
    const component = create(<Lockscreen {...props} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('handles redirect', () => {
    const component = create(<Lockscreen {...props} />);
    const div = component.root.children[0];

    expect(props.history.push).not.toHaveBeenCalled();

    act(() => {
      div.props.onTransitionEnd();
    });

    expect(props.history.push).toHaveBeenCalledWith('/home');
  });

  it('handles unlock', () => {
    const component = create(<Lockscreen {...props} />);
    const toggleSwitch = component.root.findByType(ToggleSwitch);

    act(() => {
      toggleSwitch.props.onUpdate();
    });

    expect(component.toJSON()).toMatchSnapshot();
  });
});
