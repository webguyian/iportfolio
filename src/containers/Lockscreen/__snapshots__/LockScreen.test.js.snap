// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lockscreen /> handles unlock 1`] = `
<div
  className="iportfolio-lockscreen iportfolio-lockscreen--unlocked"
  onTransitionEnd={[Function]}
>
  <header
    className="iportfolio-lockscreen-header"
  >
    <i
      className="ui-icon ui-icon-lock fas fa-lock fa-2x"
    />
    <h1
      className="ui-text ui-text--body ui-clock"
    >
      11:11
    </h1>
    <h1
      className="ui-text ui-text--body ui-date"
    >
      Tuesday, October 1
    </h1>
  </header>
  <label
    className="ui-toggle-switch"
  >
    <input
      checked={false}
      className="ui-toggle-switch-checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <button
      className="ui-btn ui-toggle-switch-slider"
      onTransitionEnd={[Function]}
      type="button"
    />
    <span
      className="ui-text ui-text--body ui-toggle-switch-label"
    >
      slide to unlock
    </span>
  </label>
</div>
`;

exports[`<Lockscreen /> renders correctly 1`] = `
<div
  className="iportfolio-lockscreen"
  onTransitionEnd={[Function]}
>
  <header
    className="iportfolio-lockscreen-header"
  >
    <i
      className="ui-icon ui-icon-lock fas fa-lock fa-2x"
    />
    <h1
      className="ui-text ui-text--body ui-clock"
    >
      11:11
    </h1>
    <h1
      className="ui-text ui-text--body ui-date"
    >
      Tuesday, October 1
    </h1>
  </header>
  <label
    className="ui-toggle-switch"
  >
    <input
      checked={false}
      className="ui-toggle-switch-checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <button
      className="ui-btn ui-toggle-switch-slider"
      onTransitionEnd={[Function]}
      type="button"
    />
    <span
      className="ui-text ui-text--body ui-toggle-switch-label"
    >
      slide to unlock
    </span>
  </label>
</div>
`;
